dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/data/NeuralNetworkDataParser.java

```diff
@@ -83,13 +90,33 @@ public void parse() {
         final Matrix data = input.get(range(0, input.getColumns() - 2));
         final int label = (int) input.get(input.getColumns() - 2);
         final boolean isValidation = ((int) input.get(input.getColumns() - 1) == 1);
-        trainingData.add(new Pair<>(new Pair<>(data, label), isValidation));
+
+        if (isValidation) {
+          validationBatchGenerator.push(data, label);
+          if (validationBatchGenerator.check()) {
```

Comment:
I'm completely fine with having the inner class `BatchGenerator`, but I still think this code can be improved to be more concise. `check()` could be called inside `push`, and the next line `dataList.add(new Pair...)` could also be run inside `BatchGenerator` instead of here. I don't quite see why that would make the code more complicated.